### ISSUE TITLE
Suggestion: Alternative 4-Panel Layout

### ISSUE THREAD

**Issue by wanqingy** [google/neuroglancer#665]
Hi! I would like to propose an alternative 4-panel layout in Neuroglancer that may better facilitate cross-referencing three orthogonal planes alongside a 3D view.

**Suggested configuration:**
Top left: xy
Bottom left: xz view
Top right: yz view
Bottom right: 3d

**Examples:** 
[Suggested new layout](https://neuroglancer-demo.appspot.com/#!%7B%22dimensions%22:%7B%22x%22:%5B8e-9%2C%22m%22%5D%2C%22y%22:%5B8e-9%2C%22m%22%5D%2C%22z%22:%5B4e-8%2C%22m%22%5D%7D%2C%22position%22:%5B120389.8203125%2C103923.046875%2C21400.57421875%5D%2C%22crossSectionScale%22:1.2712491503213994%2C%22projectionScale%22:262144%2C%22layers%22:%5B%7B%22type%22:%22image%22%2C%22source%22:%22precomputed://https://bossdb-open-data.s3.amazonaws.com/iarpa_microns/minnie/minnie65/em%22%2C%22tab%22:%22rendering%22%2C%22shaderControls%22:%7B%22normalized%22:%7B%22range%22:%5B76%2C186%5D%7D%7D%2C%22name%22:%22em%22%7D%5D%2C%22selectedLayer%22:%7B%22visible%22:true%2C%22layer%22:%22em%22%7D%2C%22layout%22:%7B%22type%22:%22row%22%2C%22children%22:%5B%7B%22type%22:%22column%22%2C%22children%22:%5B%7B%22type%22:%22viewer%22%2C%22layers%22:%5B%22em%22%5D%2C%22layout%22:%22xy%22%2C%22flex%22:1.22%7D%2C%7B%22type%22:%22viewer%22%2C%22layers%22:%5B%22em%22%5D%2C%22layout%22:%22xz%22%2C%22flex%22:0.78%7D%5D%2C%22flex%22:1.21%7D%2C%7B%22type%22:%22column%22%2C%22children%22:%5B%7B%22type%22:%22viewer%22%2C%22layers%22:%5B%22em%22%5D%2C%22layout%22:%22yz%22%2C%22flex%22:1.22%7D%2C%7B%22type%22:%22viewer%22%2C%22layers%22:%5B%22em%22%5D%2C%22layout%22:%223d%22%2C%22flex%22:0.78%7D%5D%2C%22flex%22:0.79%7D%5D%7D%7D)

![Image](https://github.com/user-attachments/assets/ee9724fc-af9c-4f27-bfe1-6cb9853dea75)

[Current default 4 panel view as an comparison](https://neuroglancer-demo.appspot.com/#!%7B%22dimensions%22:%7B%22x%22:%5B8e-9%2C%22m%22%5D%2C%22y%22:%5B8e-9%2C%22m%22%5D%2C%22z%22:%5B4e-8%2C%22m%22%5D%7D%2C%22position%22:%5B120368.4609375%2C103929.4453125%2C21376.943359375%5D%2C%22crossSectionScale%22:1.2712491503213994%2C%22projectionScale%22:262144%2C%22layers%22:%5B%7B%22type%22:%22image%22%2C%22source%22:%22precomputed://https://bossdb-open-data.s3.amazonaws.com/iarpa_microns/minnie/minnie65/em%22%2C%22tab%22:%22rendering%22%2C%22shaderControls%22:%7B%22normalized%22:%7B%22range%22:%5B76%2C186%5D%7D%7D%2C%22name%22:%22em%22%7D%5D%2C%22selectedLayer%22:%7B%22visible%22:true%2C%22layer%22:%22em%22%7D%2C%22layout%22:%224panel%22%7D)

![Image](https://github.com/user-attachments/assets/558ff8a2-85b2-4c69-91b4-aa52fdd9b3be)

**Implementation**
To experiment with this layout, I customized the layout in JSON state as follows:
```
  "layout": {
    "type": "row",
    "children": [
      {
        "type": "column",
        "children": [
          {
            "type": "viewer",
            "layers": [
              "em"
            ],
            "layout": "xy",
            "flex": 1.22
          },
          {
            "type": "viewer",
            "layers": [
              "em"
            ],
            "layout": "xz",
            "flex": 0.78
          }
        ],
        "flex": 1.21
      },
      {
        "type": "column",
        "children": [
          {
            "type": "viewer",
            "layers": [
              "em"
            ],
            "layout": "yz",
            "flex": 1.22
          },
          {
            "type": "viewer",
            "layers": [
              "em"
            ],
            "layout": "3d",
            "flex": 0.78
          }
        ],
        "flex": 0.79
      }
    ]
  }
```

In addition, I also [experimented](https://github.com/wanqingy/neuroglancer/tree/new_4panel) editing `data_panel_layout.ts` to make the suggested 4 panel as a default layout. 

**Panel Resizing:**
Another limitation of the current 4-panel layout (and 2 panel layout too) is the inability to adjust panel sizes. Using the JSON state above, users also can adjust panel sizes. 

**In summary,** it would be great if this alternative 4-panel layout can be added (if not replacing the current one). It is even better if users can rearrange and resize panels whatever they want.

**Comment by ceesem** [comment]
Having seen the demo of this yesterday, I just wanted to chime in and say that this proposed layout is awesome. It makes a lot more intuitive sense than the existing 4-panel layout (and I believe is more standard in the microscopy community), plus the ability to resize the orthogonal views to be shallower than the transverse view allows them to complement anisotropic data without taking as much real estate from the highest resolution view.

**Comment by jbms** [comment]
I'd be happy to accept a PR that adds support for resizing and/or changing the arrangement of panels.

The key question as far as supporting other arrangements is whether to allow arbitrary arrangements (e.g. maybe the user wants xy xz and 3d in one row) or just to add this one additional specific arrangement, and for either option, how to expose it in the UI.  I would indeed think that it would be better to support arbitrary arrangements.

Just modifying the existing 4panel view is not a viable option because it would affect existing Neuroglancer states/URLs.

**Comment by chrisj** [comment]
Would it be enough to just add a new name for the 4panel layout and default to the new one?

If you wanted to ensure old states continue to use the old layout even when swapping between layouts, you could have some indicating in newly generated (blank) states indicating this is a new epoch.

**Comment by jbms** [comment]
> Would it be enough to just add a new name for the 4panel layout and default to the new one?
> 
> If you wanted to ensure old states continue to use the old layout even when swapping between layouts, you could have some indicating in newly generated (blank) states indicating this is a new epoch.

Yes I think that approach would work fine.

To be sure that I'm understanding correctly, the appeal of the new layout is that the x axis is aligned for the left two panels and the y axis is aligned for the top two panels?

**Comment by wanqingy** [comment]
> To be sure that I'm understanding correctly, the appeal of the new layout is that the x axis is aligned for the left two panels and the y axis is aligned for the top two panels?

Yes!

**Comment by wanqingy** [comment]
Thank you, Jeremy!  @jbms
